ferrocene/doc/requirements/src/self-test.rst

```diff
@@ -0,0 +1,264 @@
+.. SPDX-License-Identifier: MIT OR Apache-2.0
+   SPDX-FileCopyrightText: The Ferrocene Developers
+
+``ferrocene-self-test`` tool
+============================
+
+.. req:: Purpose
+   :id: REQ_LSGKACM
+
+   The tool is used to ensure that the Ferrocene toolchain is installed correctly.
+
+.. req:: Environment
+   :id: REQ_THXJWWD
+
+   The tool should be executed on any of the host platforms listed in
+   :doc:`user-manual:targets/index`.
+
+.. req:: Execution
+   :id: REQ_P2OXII1
+
+   The tool should be executed on the command line.
+
+.. req:: Command line interface
+   :id: REQ_9A6740M
+
+   The tool accepts no command line arguments, and will ignore them if present.
+
+.. req:: Output
+   :id: REQ_42QU5MK
+
+   The tool displays checks it performs, and quits with a fail exit status when any check fails.
+
+Checks for executables
+----------------------
+
+.. req::
+   :id: REQ_0OAUY3U
+
+   These executables should be checked:
+
+   - ``rustc``
+   - ``rustdoc``
+   - ``cargo`` (if installed)
+
+.. req::
+   :id: REQ_N1VBW46
+   :implements: REQ_0OAUY3U
+
+   The executable should exist in the ``bin`` directory, relative to the root of the installation.
+
+.. req::
+   :id: REQ_R2UQ8D3
+   :implements: REQ_0OAUY3U
+
+   The executable should be a regular file.
+
+.. req::
+   :id: REQ_NUP1G0D
+   :implements: REQ_0OAUY3U
+
+   The executable should have read and execute permissions set for all users.
+
+.. req::
+   :id: REQ_GVLWOTQ
+   :implements: REQ_0OAUY3U
+
+   Check the behavior of ``--verbose --version`` command-line options.
+
+.. req::
+   :id: REQ_6OAFM70
+   :implements: REQ_GVLWOTQ
+
+   The executable should exit successfully.
+
+.. req::
+   :id: REQ_ABPRHHQ
+   :implements: REQ_GVLWOTQ
+
+   The output of the executable should be UTF-8 text.
+
+.. req::
+   :id: REQ_SL5USTK
+   :implements: REQ_GVLWOTQ
+
+   The output of the executable should have at least 3 key-value pairs,
+   with these 3 keys: ``host``, ``commit-hash``, and ``release``:
+
+   - ``host`` is the target triple of the platform that the executable is built for
+   - ``commit-hash`` is an identifier for the Git commit from which the Ferrocene release was built
+   - ``release`` is the name of the upstream release from which the Ferrocene release was based
+
+Checks for targets
+------------------
+
+.. note::
+
+   In the subsections below, we have these substitutions:
+
+   - ``$target`` refers to any of the targets listed
+   - ``$hash`` is a 16 character hexadecimal string
+
+All targets
+^^^^^^^^^^^
+
+.. note::
+
+   These checks are for all locally-installed targets listed in :doc:`user-manual:targets/index`.
```

Comment:
Should that be a requirement as well? Something like: The tool discovers all targets intalled in /path/to/targets, and executes following checks for them?

Reply:
hm, not sure... I do like the suggestion though